File: API/Controller/Controller.go
package controller

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"strings"

	_ "github.com/go-sql-driver/mysql"

	"github.com/gofiber/fiber/v2"
)

type Controller struct {
	DB  *sql.DB
	err error
}

func NewController() *Controller {
	ctrl := &Controller{}
	ctrl.connect()
	return ctrl
}

func (c *Controller) connect() bool {
	c.DB, c.err = sql.Open("mysql", "root:for00nite@tcp(127.0.0.1:3306)/BD1P1")
	if c.err != nil {
		return false
	}

	c.err = c.DB.Ping()
	if c.err != nil {
		return false
	}
	return true
}

func (c *Controller) Running(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Interpreter is running!!!",
	})
}

func (c *Controller) Test(ctx *fiber.Ctx) error {
	rows, err := c.DB.Query("SELECT * FROM Test;")
	if err != nil {
		log.Fatal(err)
	}
	defer rows.Close()

	response := []fiber.Map{}

	for rows.Next() {
		var id int
		var texto string
		if err := rows.Scan(&id, &texto); err != nil {
			log.Fatal(err)
		}
		response = append(response, fiber.Map{"id": id, "texto": texto})
	}
	if err := rows.Err(); err != nil {
		log.Fatal(err)
	}
	return ctx.JSON(fiber.Map{
		"status":   "Finish Test",
		"response": response,
	})
}

func (c *Controller) Query1(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query1",
	})
}

func (c *Controller) Query2(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query2",
	})
}

func (c *Controller) Query3(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query3",
	})
}

func (c *Controller) Query4(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query4",
	})
}

func (c *Controller) Query5(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query5",
	})
}

func (c *Controller) Query6(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query6",
	})
}

func (c *Controller) Query7(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query7",
	})
}

func (c *Controller) Query8(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query8",
	})
}

func (c *Controller) Query9(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query9",
	})
}

func (c *Controller) Query10(ctx *fiber.Ctx) error {
	return ctx.JSON(fiber.Map{
		"status": "Query10",
	})
}

func (c *Controller) Deletemodel(ctx *fiber.Ctx) error {
	script := `DROP TABLE IF EXISTS orden;
	DROP TABLE IF EXISTS producto;
	DROP TABLE IF EXISTS vendedor;
	DROP TABLE IF EXISTS cliente;
	DROP TABLE IF EXISTS categoria;
	DROP TABLE IF EXISTS pais;`

	for _, query := range strings.Split(script, ";") {
		c.DB.Exec(query)
	}

	return ctx.JSON(fiber.Map{
		"status": "Modelo eliminado",
	})
}

func (c *Controller) Createmodel(ctx *fiber.Ctx) error {
	fileData, _ := ioutil.ReadFile("../../Script/BD1P1.sql")

	for _, query := range strings.Split(string(fileData), ";") {
		c.DB.Exec(query)
	}

	return ctx.JSON(fiber.Map{
		"status": "Modelo creado",
	})
}

func (c *Controller) Deleteinfo(ctx *fiber.Ctx) error {
	script := `SET FOREIGN_KEY_CHECKS = 0;
	TRUNCATE TABLE orden;
	TRUNCATE TABLE producto;
	TRUNCATE TABLE vendedor;
	TRUNCATE TABLE cliente;
	TRUNCATE TABLE categoria;
	TRUNCATE TABLE pais;
	SET FOREIGN_KEY_CHECKS = 1;`

	for _, query := range strings.Split(script, ";") {
		c.DB.Exec(query)
	}

	return ctx.JSON(fiber.Map{
		"status": "Informacion eliminada",
	})
}

func (c *Controller) Postdatamodel(ctx *fiber.Ctx) error {
	files := []string{
		"../../Data/paises.csv",
		"../../Data/Categorias.csv",
		"../../Data/clientes.csv",
		"../../Data/vendedores.csv",
		"../../Data/productos.csv",
		"../../Data/ordenes.csv",
	}

	for i, r := range files {
		file, err := os.Open(r)
		if err != nil {
			log.Fatal(err)
		}
		defer file.Close()

		reader := csv.NewReader(file)
		reader.Comma = ';'
		header := true

		for {
			line, err := reader.Read()
			if err == io.EOF {
				break
			}
			if err != nil {
				fmt.Println("\033[96mNo se pudo leer la linea del csv\033[0m")
				continue
			}
			if header {
				header = false
				continue
			}
			if i == 0 { // Pais
				_, err := c.DB.Query("INSERT INTO pais (id, nombre) VALUE (?, ?)", line[0], line[1])
				if err != nil {
					fmt.Println("Error al insertar pais", err)
				}
			} else if i == 1 { // Categoria
				_, err := c.DB.Query("INSERT INTO categoria (id, nombre) VALUE (?, ?)", line[0], line[1])
				if err != nil {
					fmt.Println("Error al insertar categoria", err)
				}
			} else if i == 2 { // Cliente
				_, err := c.DB.Query("INSERT INTO cliente (id, nombre, apellido, direccion, telefono, tarjeta, edad, salario, genero, pais_id) VALUE (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)", line[0], line[1], line[2], line[3], line[4], line[5], line[6], line[7], line[8], line[9])
				if err != nil {
					fmt.Println("Error al insertar cliente", err)
				}
			} else if i == 3 { // Vendedor
				fullName := strings.Split(line[1], " ")
				_, err := c.DB.Query("INSERT INTO vendedor (id, nombre, apellido, pais_id) VALUE (?, ?, ?, ?)", line[0], fullName[0], fullName[1], line[2])
				if err != nil {
					fmt.Println("Error al insertar vendedor", err)
				}
			} else if i == 4 { // Producto
				_, err := c.DB.Query("INSERT INTO producto (id, nombre, precio, categoria_id) VALUE (?, ?, ?, ?)", line[0], line[1], line[2], line[3])
				if err != nil {
					fmt.Println("Error al insertar producto", err)
				}
			} else { // Orden
				date := strings.Split(line[2], "/")
				line[2] = fmt.Sprintf("%s-%s-%s", date[2], date[1], date[0])
				_, err := c.DB.Query("INSERT INTO orden (id, linea, fecha, cantidad, cliente_id, vendedor_id, producto_id) VALUE (?, ?, ?, ?, ?, ?, ?)", line[0], line[1], line[2], line[6], line[3], line[4], line[5])
				if err != nil {
					fmt.Println("Error al insertar orden", err)
				}
			}
		}
	}

	return ctx.JSON(fiber.Map{
		"status": "Informacion insertada",
	})
}

File: API/Python/Controller.py
import mysql.connector
from dotenv import load_dotenv
import csv
import os

# Carga las variables de entorno desde el archivo .env
load_dotenv()

class Controller:
    def __init__(self) -> None:
        self.conexion = mysql.connector.connect(
            user=os.getenv('DB_USER'),
            password=os.getenv('DB_PASS'),
            host=os.getenv('DB_HOST'),
            port=os.getenv('DB_PORT'),
            database=os.getenv('DB_NAME'),
        )
        print(os.getenv('DB_USER'))
        self.cursor = self.conexion.cursor()

    def deletemodel(self):
        try:
            query = f'''DROP TABLE IF EXISTS orden;
            DROP TABLE IF EXISTS datoorden;
            DROP TABLE IF EXISTS producto;
            DROP TABLE IF EXISTS vendedor;
            DROP TABLE IF EXISTS cliente;
            DROP TABLE IF EXISTS categoria;
            DROP TABLE IF EXISTS pais;'''
            for q in query.split('\n'):
                self.cursor.execute(q)
            return {'status': 'Modelo eliminado'}, 200
        except Exception as e:
            print(e)
            return {'status': 'Error al eliminar modelo'}, 500

    def createmodel(self):
        try:
            query = open('../../Script/BD1P1.sql').read()
            for q in query.split(';'):
                self.cursor.execute(q)
            return {'status': 'Modelo creado'}, 200
        except Exception as e:
            print(e)
            return {'status': 'Error al crear modelo'}, 500

    def deleteinfo(self):
        try:
            query = f'''SET FOREIGN_KEY_CHECKS = 0;
            TRUNCATE TABLE orden;
            TRUNCATE TABLE datoorden;
            TRUNCATE TABLE producto;
            TRUNCATE TABLE vendedor;
            TRUNCATE TABLE cliente;
            TRUNCATE TABLE categoria;
            TRUNCATE TABLE pais;
            SET FOREIGN_KEY_CHECKS = 1;'''
            for q in query.split('\n'):
                self.cursor.execute(q)
            return {'status': 'Informacion eliminada'}, 200
        except Exception as e:
            print(e)
            return {'status': 'Error al eliminar informacion'}, 500

    def postdatamodel(self):
        try:
            files = {
                0: '../../Data/paises.csv',
                1: '../../Data/Categorias.csv',
                2: '../../Data/clientes.csv',
                3: '../../Data/vendedores.csv',
                4: '../../Data/productos.csv',
                5: '../../Data/ordenes.csv',
            }

            for i, r in files.items():
                file = open(r, newline='', encoding='utf-8')
                reader = csv.reader(file, delimiter=';')
                next(reader, None)

                for line in reader:
                    if i == 0:
                        self.cursor.execute(f'''INSERT INTO pais (id, nombre) VALUE ({line[0]}, "{line[1]}")''')
                        self.conexion.commit()
                    elif i == 1:
                        self.cursor.execute(f'''INSERT INTO categoria (id, nombre) VALUE ({line[0]}, "{line[1]}")''')
                        self.conexion.commit()
                    elif i == 2:
                        self.cursor.execute(f'''INSERT INTO cliente (id, nombre, apellido, direccion, telefono, tarjeta, edad, salario, genero, pais_id) VALUE ({line[0]}, "{line[1]}", "{line[2]}", "{line[3]}", "{line[4]}", "{line[5]}", {line[6]}, {line[7]}, "{line[8]}", {line[9]})''')
                        self.conexion.commit()
                    elif i == 3:
                        fullName = line[1].split(" ")
                        self.cursor.execute(f'''INSERT INTO vendedor (id, nombre, apellido, pais_id) VALUE ({line[0]}, "{fullName[0]}", "{fullName[1]}", {line[2]})''')
                        self.conexion.commit()
                    elif i == 4:
                        self.cursor.execute(f'''INSERT INTO producto (id, nombre, precio, categoria_id) VALUE ({line[0]}, "{line[1]}", {line[2]}, {line[3]})''')
                        self.conexion.commit()
                    else:
                        date = line[2].split("/")
                        line[2] = f"{date[2]}-{date[1]}-{date[0]}"
                        self.cursor.execute(f'''INSERT INTO orden (id, fecha, cliente_id) SELECT {line[0]}, '{line[2]}', {line[3]} FROM orden WHERE id != {line[0]}''')
                        self.cursor.execute(f'''INSERT INTO datoorden (linea, cantidad, orden_id, vendedor_id, producto_id) VALUE ({line[1]}, {line[6]}, {line[0]}, {line[4]}, {line[5]})''')
                        self.conexion.commit()

            return {'status': 'Informacion insertada'}, 200
        except Exception as e:
            print(e)
            return {'status': 'Error al insertar informacion'}, 500